CI Feature to detect stalled experiments

TerrenceMcGuinness-NOAA opened the pull request:
# Description

A helper python script `rocoto_statcount.py` is added in `$HOMEgfs/ci/scrpts/utils` that checks to see if an experiment is "stalled".   This is done by running **rocotorun** and checking within all the jobs and if none are RUNNING, SUBMITTING, or QUEUED then the experiment is STALLED.

 Resolves #2008 


# Type of change
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
- Ran in CI CRON development space and set up a case that had a missing file dependency and it was appropriately flagged.

# Checklist
- [ ] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

TerrenceMcGuinness-NOAA committed:
started to add python rocoto stat

TerrenceMcGuinness-NOAA committed:
first steps with rocoto stat script

TerrenceMcGuinness-NOAA committed:
finished working version of rocoto_statcount

TerrenceMcGuinness-NOAA committed:
replaced explicit decleration of stat dict with loop in rocoto statcount

TerrenceMcGuinness-NOAA committed:
fixed end state mixup of varialble names

TerrenceMcGuinness-NOAA committed:
more slight improvments on user GitHub messaging outputs

TerrenceMcGuinness-NOAA committed:
added failed state updates with stalled in run ci

TerrenceMcGuinness-NOAA committed:
Update run_ci.sh

quotes from bash linter

TerrenceMcGuinness-NOAA committed:
Update run_ci.sh

jobs should not be plurl

TerrenceMcGuinness-NOAA committed:
bug updates after testing depedancies catch

TerrenceMcGuinness-NOAA committed:
fixed conflict with run_ci

aerorahul commented on ci/scripts/check_ci.sh:
```diff
@@ -131,7 +131,7 @@ for pr in ${pr_list}; do
       error_logs=$("${rocotostat}" -d "${db}" -w "${xml}" | grep -E 'FAIL|DEAD' | awk '{print "-c", $1, "-t", $2}' | xargs "${rocotocheck}" -d "${db}" -w "${xml}" | grep join | awk '{print $2}') || true
       {
-       echo "Experiment ${pslot} Terminated: *** FAILED ***"
+       echo "Experiment ${pslot} Terminated: *** FAILED *** on ${MACHIND_ID^}"
```
Small typo
```suggestion
       echo "Experiment ${pslot} Terminated: *** FAILED *** on ${MACHINE_ID^}"
```

TerrenceMcGuinness-NOAA committed:
Update rocoto_statcount.py

fixed some grammar in description

aerorahul commented on ci/scripts/check_ci.sh:
```diff
@@ -152,8 +152,8 @@ for pr in ${pr_list}; do
       DATE=$(date)
-      echo "Experiment ${pslot} **SUCCESS** ${DATE}" >> "${output_ci_single}"
-      echo "Experiment ${pslot} **SUCCESS** at ${DATE}" >> "${output_ci}"
+      echo "Experiment ${pslot} **SUCCESS** ${DATE +%Y%m%d} on ${MACHINE_ID^}" >> "${output_ci_single}"
+      echo "Experiment ${pslot} **SUCCESS** at ${DATE +%Y%m%d} on ${MACHIND_ID^}" >> "${output_ci}"
```
This cannot possibly work `${DATE +%Y%m%d}`.

TerrenceMcGuinness-NOAA committed:
Update run_ci.sh

fixed double quotes for bash lint

aerorahul commented on ci/scripts/run_ci.sh:
```diff
@@ -81,7 +83,20 @@ for pr in ${pr_list}; do
+      output_ci="${pr_dir}/output_runtime_single.log"
+      {
+        echo "${pslot} has *** STALLED **** on ${MACHINE_ID^}"
+        echo "A job in expermint ${pslot} in ${pslot_dir}"
```
```suggestion
        echo "A job in experiment ${pslot} in ${pslot_dir}"
```

aerorahul reviewed (commented):
looks ok, but I spotted some minor typos.
See suggestions.
And one comment on something that probably is incorrect.

WalterKolczynski-NOAA commented on ci/scripts/utils/rocoto_statcount.py:
```diff
@@ -0,0 +1,104 @@
+       'Done_Cycles' :  sum([ sublist.count('Done') for sublist in rocotostat_output ])
+    }
+    
+    status_cases = [ 'SUCCEEDED', 'FAIL', 'DEAD', 'RUNNING', 'PENDING', 'QUEUED']
```
Is 'PENDING' a valid status, or do you mean 'SUBMITTED' (which needs one).

WalterKolczynski-NOAA replied:
Oh, and 'UNKNOWN' is also a valid state, and should either be automatically rewound or treated like a fail. It happens when rocoto is unable to get information about a job from the scheduler (usually because it wasn't run for a while and the job aged off, like during a long downtime).

TerrenceMcGuinness-NOAA replied:
Its actually **SUBMITTING** (which you got right later in this review) but yes you are correct there is no **PENDING**

And yeah, I agree we should add **UNKOWN** to **FAILED** since _Slurm_ would be "off" in that case and the experiment at that point would be questionable.

```
tmcguinness ~/NOAA-EMC/rocoto (develop) $ find . -type f  | xargs grep -h  ":state]=" | grep SUBMIT
            next unless @active_jobs[taskname][cycle][:state]=="SUBMITTING"
          newjob[:state]="SUBMITTING"
tmcguinness ~/NOAA-EMC/rocoto (develop) $ find . -type f  | xargs grep -h  ":state]=" | grep PEND
tmcguinness ~/NOAA-EMC/rocoto (develop) $ 
```

Looks like **FAIL** should be **FAILED** too:
```
tmcguinness ~/NOAA-EMC/rocoto (develop) $ find . -type f  | xargs grep -h  "record\[\:state\]=" | grep -Ev 'if|unless' | sed 's/ //g' | cut -d"=" -f2 | sed 's/"//g' | sed "s/'//g" | sort | uniq 
FAILED
QUEUED
RUNNING
SUCCEEDED
UNKNOWN
USERHOLD
```

WalterKolczynski-NOAA commented on ci/scripts/utils/rocoto_statcount.py:
```diff
@@ -0,0 +1,104 @@
+    rocoto_status = rocoto_statcount()
+    for status in rocoto_status:
+        print(f'Number of {status} : {rocoto_status[status]}')
+    rocoto_state = 'Running'
+    if rocoto_status['Cycles'] == rocoto_status['Done_Cycles']:
+        rocoto_state = 'Done'
+
+    if args.check_stalled:
+        if rocoto_state != 'Done':
+            rocoto_run = which("rocotorun")
+            rocoto_run.add_default_arg(['-w',args.w,'-d',args.d])
+            rocoto_run()
+            rocoto_status2 = rocoto_statcount()
+            if rocoto_status2 == rocoto_status:
+                rocoto_state = 'Stalled'
+                print(f'Rocoto State : {rocoto_state}')
+                sys.exit(-1)
+            else:
+                rocoto_state = 'Running'
```
I don't think this is robust. It's not guaranteed the statuses to change on a `rocotorun` (particularly if a long job is running). As we discussed before, what we should be looking for is the number of successful or pending/queued jobs. You don't even need to call `rocotorun` the second time or have this extra `check_stalled` option.
```suggestion
    if rocoto_status['Cycles'] == rocoto_status['Done_Cycles']:
        rocoto_state = 'Done'
    elif rocoto_status['RUNNING'] + rocoto_status['SUBMITTING'] + rocoto_status['QUEUED'] == 0:
        rocoto_state = 'Stalled'
    else:
        rocoto_state = 'Running'
```

aerorahul replied:
`elif any(x in rocoto_status for x in ['RUNNING', 'SUBMITTING', 'QUEUED'])` should do for line 90.
@aerorahul not exactly, we need to check that they are all **_not_** RUNNING, SUBMITTING, or QUEUED as Walt has the total SUM be zero as the eval in dict has number of occurrences.

TerrenceMcGuinness-NOAA replied:
Oh right @WalterKolczynski-NOAA of course, not advancing doesn't mean the all the status states don't change, its when after  running **rocotorun** there are none that are RUNNING, SUBMITTING, or QUEUED.  Really simplifies things indeed.

aerorahul commented on ci/scripts/utils/rocoto_statcount.py:
```diff
@@ -0,0 +1,104 @@
+import sys
+import os
+
+from wxflow import Executable, which, Logger
```
```suggestion
from wxflow import Executable, which, Logger, CommandNotFoundError
```

aerorahul commented on ci/scripts/utils/rocoto_statcount.py:
```diff
@@ -0,0 +1,104 @@
+
+    args = input_args()
+
+    rocotostat = which("rocotostat")
+    if not rocotostat:
+        logger.exception("rocotostat not found in PATH")
+        sys.exit(-1)
```
```suggestion
    rocotostat = which("rocotostat")
    if not rocotostat:
        logger.exception("rocotostat not found in PATH")
        sys.exit(-1)
    try:
        rocotostat = which("rocotostat")
    except CommandNotFoundError:
        logger.exception("rocotostat not found in PATH")
        raise CommandNotFoundError("rocotostat not found in PATH")
```

aerorahul commented on ci/scripts/utils/rocoto_statcount.py:
```diff
@@ -0,0 +1,104 @@
+    xml_file_path = os.path.abspath(args.w)
+    db_file_path = os.path.abspath(args.d)
+
+    rocotostat_all = which("rocotostat")
```
Why do this again?  You already have `rocotostat`.  You can use that any number of times with different arguments.

@aerorahul I have not been able to find a way to use `add_default_args` to "update" the argument list of an Executable Object.  Can you fine an example, could not find such a use case in the repo.

TerrenceMcGuinness-NOAA replied:
Could not find a way to "change" arguments once the are set on the Executable object.

aerorahul commented on ci/scripts/utils/rocoto_statcount.py:
```diff
@@ -0,0 +1,104 @@
+    db_file_path = os.path.abspath(args.d)
+
+    rocotostat_all = which("rocotostat")
+    rocotostat.add_default_arg(['-w',xml_file_path,'-d',db_file_path,'-s'])
+    rocotostat_all.add_default_arg(['-w',xml_file_path,'-d',db_file_path,'-a'])
+
+    rocotostat_output = rocotostat(output=str)
+    rocotostat_output = rocotostat_output.splitlines()[1:]
+    rocotostat_output = [line.split()[0:2] for line in rocotostat_output]
+
+    rocotostat_output_all = rocotostat_all(output=str)
+    rocotostat_output_all = rocotostat_output_all.splitlines()[1:]
+    rocotostat_output_all = [line.split()[0:4] for line in rocotostat_output_all]
+    rocotostat_output_all = [line for line in rocotostat_output_all if len(line) != 1]
```
I can show you how to reduce this and simplify it considerably

TerrenceMcGuinness-NOAA replied:
Sure that would be great, you can jot down pseudo code it you like

aerorahul commented on ci/scripts/utils/rocoto_statcount.py:
```diff
@@ -0,0 +1,104 @@
+
+    if args.check_stalled:
+        if rocoto_state != 'Done':
+            rocoto_run = which("rocotorun")
```
Having seen this, I think we should have a `RocotoCommands` class and initialize it with everything from rocoto you need; e.g. `rocotostat`, `rocotorun`, etc.  While you are at it, also provide the `xml` and `database` to the class object rather than have to do this over and over.

aerorahul reviewed (commented):
some suggestions for `rocoto_statcount.py`

TerrenceMcGuinness-NOAA committed:
spelling, typos, and brain fart on {DATE +}

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/rocoto_statcount.py

update to more robust logic

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/rocoto_statcount.py

added correct exception when command is not found

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/rocoto_statcount.py

added import for correct exception

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
updated elseif line and replaced PENDING with SUBMITTING

TerrenceMcGuinness-NOAA committed:
added exit code and removed redundent which

TerrenceMcGuinness-NOAA committed:
put Walters test back for no RUNNING, SUBMITTING, or QUEUED and added UNKNOWN as a final state

TerrenceMcGuinness-NOAA committed:
added back which method to get new and sperate Execute Object to have alternative argument list to rocotostat

TerrenceMcGuinness-NOAA committed:
Update check_ci.sh

or true for echo date commands

TerrenceMcGuinness-NOAA committed:
Update run_ci.sh

syntax fix on quote in rocotorun command line

TerrenceMcGuinness-NOAA committed:
Update run_ci.sh

removed old switch for check stalled that has been depricated

TerrenceMcGuinness-NOAA committed:
removed DATE as date command

TerrenceMcGuinness-NOAA committed:
white spaces from lint

TerrenceMcGuinness-NOAA committed:
pynorm indent on dict

TerrenceMcGuinness-NOAA committed:
more white space related pynorm stuff

TerrenceMcGuinness-NOAA committed:
hopefully last white space related pynorm stuff

TerrenceMcGuinness-NOAA committed:
Update run_ci.sh

trying to force refresh

TerrenceMcGuinness-NOAA committed:
Update run_ci.sh

stabs at solving double quote requirements for shellnoms

TerrenceMcGuinness-NOAA committed:
Update run_ci.sh

double quoting flags wont stop even though the are correct

TerrenceMcGuinness-NOAA committed:
Update run_ci.sh

still wrestling with double quote complaints from shellnormns even thought they are correct in the script and trying that ever the F&*@

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/detect_dependency

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/detect_dependency

TerrenceMcGuinness-NOAA committed:
updated discription for rocoto_statcount

TerrenceMcGuinness-NOAA reviewed (commented):
Done with my review

WalterKolczynski-NOAA commented on ci/scripts/run_ci.sh:
```diff
@@ -81,7 +83,22 @@ for pr in ${pr_list}; do
-    echo "Running: ${rocotorun} -v 10 -w ${xml} -d ${db}"
     "${rocotorun}" -v 10 -w "${xml}" -d "${db}"
+    set +e
+    "${ROOT_DIR}"/ci/scripts/utils/rocoto_statcount.py -d "${db}" -w "${xml}"
```
```suggestion
    "${ROOT_DIR}/ci/scripts/utils/rocoto_statcount.py" -d "${db}" -w "${xml}"
```

WalterKolczynski-NOAA commented on ci/scripts/run_ci.sh:
```diff
@@ -81,7 +83,22 @@ for pr in ${pr_list}; do
+    rc=$?
+    
+    if [[ "${rc}" -ne 0 ]]; then
+      output_ci="${pr_dir}"/output_runtime_single.log
```
```suggestion
      output_ci="${pr_dir}/output_runtime_single.log"
```

WalterKolczynski-NOAA commented on ci/scripts/utils/rocoto_statcount.py:
```diff
@@ -0,0 +1,99 @@
+        The parsed command-line arguments.
+    """
+
+    description = """
+        Using rocotostat to get the status of all jobs this scripts
+        determines rocoto_state: if all cycles are done, then rocoto_state is Done.
+        Assuming rocotorun had just been run, and the rocoto_state is not Done, then
+        rocoto_state is Stalled if there are no jobs that are RUNNING, SUBMITTING, or QUEUED.
+        """
```
This seems more like a description for the whole script or `rocoto_statcount` than this function.

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/detect_dependency

WalterKolczynski-NOAA commented:
@TerryMcGuinness-NOAA still waiting on you to make the suggestions I recommended

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/run_ci.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/run_ci.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>
